### ISSUE TITLE
ResizeObserver#observeComponentChanges throws unhandleable RuntimeException

### ISSUE THREAD

**Issue by marcobsidian** [viritin/flow-viritin#75]
Hi Viritin Team,

We recently added this library to our project for its `ResizeObserver` and it's already a huge improvement over the `BrowserWindowResizeListener` we had to use previously. However, there is currently one issue that we can't really work around.

Current issue
-
Registering components to the `ResizeObserver` at the right time is tricky, as we are using custom components that are dynamically attached to the UI based on user interaction (e.g. when a Tab is selected that should display them).

If we are registering them too early (e.g. as part of the constructor) and the component isn't attached at initial page load, a `RuntimeException` is thrown from deep within Vaadin via the following JavaScript-Execution:

https://github.com/viritin/flow-viritin/blob/6e592a812f89b9ea58ab826006e1e1c9728d3e56/src/main/java/org/vaadin/firitin/util/ResizeObserver.java#L110-L115

This then triggers our custom error handler and displays a confusing error message to the user:

> java.lang.RuntimeException: Error adding size observer, component not attached!?

Using `Component#onAttach` is something we tried, but did not work for our use case, so we are currently forced to register the component via a delayed JavaScript-Execution. This works relatively well, but still results in the Exception being thrown if the user quickly switches between routes.

For us, getting the exception just means that calculating the size of the component failed, as it is not there anymore, which we don't really care about, since... well, it's not there anymore!

Possible solutions
-
One of the following changes to the API would allow us to use it without having to work around the errors constantly:

- [ ] Introducing a new exception type (e.g. `ResizeObserverException`) instead of throwing a generic `RuntimeException`
-> We could handle this exception type differently with our ErrorHandler (just log, no error window)
- [ ] Extracting the error handler as new method `protected void handleNotAttached(String message){...}`
-> We could introduce a new custom type that extends `ResizeObserver` and overrides the new error handler function
- [ ] Extracting the error handler as a `final SerializableConsumer<String> errorHandler` that is passed via constructor/ factory methods. The default behavior could stay the same by overloading existing factory methods.
-> We could use the new overloads `ResizeObserver.get(s -> logger.error(s))` or `ResizeObserver.of(ui, s -> logger.error(s))`

**Comment by mstahv** [comment]
Hi, could you share some somewhat realistic usage example where the issue happens? Might help to figure out the best solutions.

**Comment by mstahv** [comment]
Tried to make a test case myself, I guess this is the fundamental issue? https://github.com/viritin/flow-viritin/commit/9e0efd9c7cc0d2ab87131c5a0158e85482d0de94

**Comment by mstahv** [comment]
The basic case as in the issue is now tackled, but it might be that based on your description this is not enough for you. Would be great if you would try (I'll cut a new version soon) and if not, let's prepare a more complete test case.

If this fix works for you, please comment here and lets close the isseu.

**Comment by marcobsidian** [comment]
Sorry that I wasn't able to respond yet. Thanks a lot for tackling the issue ❤️ 

I'll try out your fix on our project tomorrow by overwriting the class file with the updated version.

If not, I will setup a better example based on what we are doing 👍

**Comment by marcobsidian** [comment]
This fix allowed us to get rid of the JavaScript workaround, which is already an improvement!

However, the error still appears regularly (but less frequently it seems like) when navigating between routes (we have a navigation bar with RouterLinks for that).

I put a breakpoint at point where the error is thrown and opened Debugger and when the error was thrown I observed the following:
![ResizeObserver](https://github.com/user-attachments/assets/842387ed-39a8-47d5-b957-39dc4cf14d5b)

What I find interesting about this:
- c.isAttached() still returns false at the point the error is thrown
- c.getElement() has two flags #hasBeenAttached and #hasBeenDetached that both are set to true
- I also tried c.getUI() and it returned an empty Optional

Maybe these findings already help you figure out what causes the error? I will try to setup a reproducable example in my private project in the meantime.

**Comment by marcobsidian** [comment]
I modified Line 116 to see if that would be the fix, but it clearly wasn't, so ignore that part 😅

**Comment by marcobsidian** [comment]
I was able to get a reliable reproduction using the following example:

```java
package com.marcobsidian.vaadin.views.resizeobserver;

import org.vaadin.firitin.util.ResizeObserver;

import com.marcobsidian.vaadin.views.MainLayout;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.Composite;
import com.vaadin.flow.component.HasComponents;
import com.vaadin.flow.component.Text;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.router.Route;

public abstract class ResizeObserverView<C extends Component & HasComponents> extends Composite<C> {

    private ResizeObserver resizeObserver;

    protected ResizeObserverView() {
        super();
        getResizeObserver().observe(getContent(), dimensions -> {
            System.out.println("Height: " + dimensions.height());
            System.out.println("Width: " + dimensions.width());
        });
    }

    @Override
    protected C initContent() {
        var content = super.initContent();
        try {
            Thread.sleep(500); // Simulate database query or longer reflective operation
            content.add(new Text("foo"));
        }
        catch (InterruptedException e) {
            e.printStackTrace();
        }
        return content;
    }

    public ResizeObserver getResizeObserver() {
        if (resizeObserver == null) {
            resizeObserver = ResizeObserver.get();
        }
        return resizeObserver;
    }

    @Route(value = "resize1", layout = MainLayout.class)
    public static class ResizeObserver1View extends ResizeObserverView<VerticalLayout> {

    }

    @Route(value = "resize2", layout = MainLayout.class)
    public static class ResizeObserver2View extends ResizeObserverView<VerticalLayout> {

    }

    @Route(value = "resize3", layout = MainLayout.class)
    public static class ResizeObserver3View extends ResizeObserverView<VerticalLayout> {

    }

}
```

The class `MainLayout` is just a regular `AppLayout` where i added the views to quickly navigate like that:
```java

    private VerticalLayout createNavigation() {
        VerticalLayout nav = new VerticalLayout();

        ...

        nav.add(new RouterLink("ResizeObserver1", ResizeObserver1View.class));
        nav.add(new RouterLink("ResizeObserver2", ResizeObserver2View.class));
        nav.add(new RouterLink("ResizeObserver3", ResizeObserver3View.class));

        return nav;
    }

```

**Comment by mstahv** [comment]
I [added these test cases](https://github.com/viritin/flow-viritin/commit/504dad72fa757b9e6a26d3d00f4ee8c5480f3417) to the project, but could not get the issue reproduced. What do you do in the UI? Is it some specific browser?

**Comment by marcobsidian** [comment]
Huh, that is weird... I am just using regular Google Chrome.

I am sending over a reproduction video

**Comment by marcobsidian** [comment]
https://github.com/user-attachments/assets/2214991e-6168-4398-9ef6-86856147dbd3

**Comment by mstahv** [comment]
Thanks, got it reproduced now! Didn't iterate the view furiously enough :-)

It looks to me that there is some edge case issue in Flow JS executions. I guess such case can simply be eaten (or logged with fine). I'll make some refactoring + implement some cleanup at the same time.

**Comment by mstahv** [comment]
Hi!

2.10.3 currently on its way to central. I believe/hope it has this isseu fixed now. I also did some refactoring and JavaDocs. Refactoring might cause tiny backwards incompatibility, if you are not using lambda expressions (which I believe most are).

I hope to push a derived version of the functionality to Vaadin as PR at some point, but when I started preparing for that, I guess the API could be much better if dropped to the core. Listeners could then be intuitivily added directly to components (and framework could delegate them to an internal helper like this).
